## Glassdoor Data

### Introduction to dataset

In [274]:
import pandas as pd
import numpy as np

In [275]:
df = pd.read_csv('Uncleaned_DS_Jobs.csv')

In [276]:
df.head()

index          Job Title               Salary Estimate  \
0      0  Sr Data Scientist  $137K-$171K (Glassdoor est.)   
1      1     Data Scientist  $137K-$171K (Glassdoor est.)   
2      2     Data Scientist  $137K-$171K (Glassdoor est.)   
3      3     Data Scientist  $137K-$171K (Glassdoor est.)   
4      4     Data Scientist  $137K-$171K (Glassdoor est.)   

                                     Job Description  Rating  \
0  Description\n\nThe Senior Data Scientist is re...     3.1   
1  Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2  Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3  JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   

              Company Name       Location            Headquarters  \
0         Healthfirst\n3.1   New York, NY            New York, NY   
1             ManTech\n4.2  Chantilly, VA             Herndon, VA   
2      Analysis Group\n3.8     Boston, MA              Boston, MA   
3             INFICON\n3.5     Newton, MA  Bad Ragaz, Switzerland   
4  Affinity Solutions\n2.9   New York, NY            New York, NY   

                      Size  Founded        Type of ownership  \
0   1001 to 5000 employees     1993   Nonprofit Organization   
1  5001 to 10000 employees     1968         Company - Public   
2   1001 to 5000 employees     1981  Private Practice / Firm   
3    501 to 1000 employees     2000         Company - Public   
4      51 to 200 employees     1998        Company - Private   

                                Industry             Sector  \
0                     Insurance Carriers          Insurance   
1                 Research & Development  Business Services   
2                             Consulting  Business Services   
3  Electrical & Electronic Manufacturing      Manufacturing   
4                Advertising & Marketing  Business Services   

                      Revenue  \
0    Unknown / Non-Applicable   
1      $1 to $2 billion (USD)   
2  $100 to $500 million (USD)   
3  $100 to $500 million (USD)   
4    Unknown / Non-Applicable   

                                         Competitors  
0            EmblemHealth, UnitedHealth Group, Aetna  
1                                                 -1  
2                                                 -1  
3  MKS Instruments, Pfeiffer Vacuum, Agilent Tech...  
4               Commerce Signals, Cardlytics, Yodlee

Interpreting a row: <br>
 The Job Senior Data Scientist from the company Healthfirst which is a Nonprofit Organization in the Insurance industry with about 1001 to 5000 employees and has the headquarter in New York, has a salary range between 137k$ to 171k$. It was founded in 1993 and has as competitors EmblemHealth, UnitedHealth Group and Aetna. 

### Summary Statistics

In [277]:
df.describe(include="all")

index       Job Title              Salary Estimate  \
count   672.000000             672                          672   
unique         NaN             172                           30   
top            NaN  Data Scientist  $79K-$131K (Glassdoor est.)   
freq           NaN             337                           32   
mean    335.500000             NaN                          NaN   
std     194.133974             NaN                          NaN   
min       0.000000             NaN                          NaN   
25%     167.750000             NaN                          NaN   
50%     335.500000             NaN                          NaN   
75%     503.250000             NaN                          NaN   
max     671.000000             NaN                          NaN   

                                          Job Description      Rating  \
count                                                 672  672.000000   
unique                                                489         NaN   
top     Job Overview: The Data Scientist is a key memb...         NaN   
freq                                                   12         NaN   
mean                                                  NaN    3.518601   
std                                                   NaN    1.410329   
min                                                   NaN   -1.000000   
25%                                                   NaN    3.300000   
50%                                                   NaN    3.800000   
75%                                                   NaN    4.300000   
max                                                   NaN    5.000000   

          Company Name           Location  Headquarters                 Size  \
count              672                672           672                  672   
unique             432                207           229                    9   
top     Hatch Data Inc  San Francisco, CA  New York, NY  51 to 200 employees   
freq                12                 69            33                  135   
mean               NaN                NaN           NaN                  NaN   
std                NaN                NaN           NaN                  NaN   
min                NaN                NaN           NaN                  NaN   
25%                NaN                NaN           NaN                  NaN   
50%                NaN                NaN           NaN                  NaN   
75%                NaN                NaN           NaN                  NaN   
max                NaN                NaN           NaN                  NaN   

            Founded  Type of ownership Industry                  Sector  \
count    672.000000                672      672                     672   
unique          NaN                 13       58                      23   
top             NaN  Company - Private       -1  Information Technology   
freq            NaN                397       71                     188   
mean    1635.529762                NaN      NaN                     NaN   
std      756.746640                NaN      NaN                     NaN   
min       -1.000000                NaN      NaN                     NaN   
25%     1917.750000                NaN      NaN                     NaN   
50%     1995.000000                NaN      NaN                     NaN   
75%     2009.000000                NaN      NaN                     NaN   
max     2019.000000                NaN      NaN                     NaN   

                         Revenue Competitors  
count                        672         672  
unique                        14         108  
top     Unknown / Non-Applicable          -1  
freq                         213         501  
mean                         NaN         NaN  
std                          NaN         NaN  
min                          NaN         NaN  
25%                          NaN         NaN  
50%                          NaN         NaN  

There's 672 job titles on the dataset with 172 different titles, but the most common one is 'Data Scientist' which appear 337 times; <br>
The salary estimate that appears the most is between 75k to 131k; <br>
The average rating is 3.52, but there's companies with a rating of 5 and we can already see an error since there are companies with rating of -1; <br>
The company that displays the most job descriptions is 'Health Data Inc': <br>
The most common location is 'San Francisco, CA' and the headquarters is 'New York, NY'; <br>
Most companies are of small size with between 51 to 200 employees and are private, belonging to the IT Sector; <br>
There is clearly an error in 'Industry' since the most common is '-1'; <br>
The year of founding ranges between 1917 and 2019;

## Data Transformation

### 1) Create salary as integer

Step 1) Remove 'K' <br>
Step 2) For each number add '000' <br>
Step 3) Remove the dollar sign $ <br>
Step 4) Remove the sources (Glassdoor est.) and (Employer est.) <br>
Step 5) Create lower band and high band splitting on '-'

In [278]:
df['Salary Estimate'] = df['Salary Estimate'].apply(lambda x: x.replace('K','000')).apply(lambda x: x.replace('$', '')).apply(lambda x: x.replace('(Glassdoor est.)','')).apply(lambda x: x.replace('(Employer est.)',''))

In [279]:
df.head()

index          Job Title Salary Estimate  \
0      0  Sr Data Scientist  137000-171000    
1      1     Data Scientist  137000-171000    
2      2     Data Scientist  137000-171000    
3      3     Data Scientist  137000-171000    
4      4     Data Scientist  137000-171000    

                                     Job Description  Rating  \
0  Description\n\nThe Senior Data Scientist is re...     3.1   
1  Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2  Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3  JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   

              Company Name       Location            Headquarters  \
0         Healthfirst\n3.1   New York, NY            New York, NY   
1             ManTech\n4.2  Chantilly, VA             Herndon, VA   
2      Analysis Group\n3.8     Boston, MA              Boston, MA   
3             INFICON\n3.5     Newton, MA  Bad Ragaz, Switzerland   
4  Affinity Solutions\n2.9   New York, NY            New York, NY   

                      Size  Founded        Type of ownership  \
0   1001 to 5000 employees     1993   Nonprofit Organization   
1  5001 to 10000 employees     1968         Company - Public   
2   1001 to 5000 employees     1981  Private Practice / Firm   
3    501 to 1000 employees     2000         Company - Public   
4      51 to 200 employees     1998        Company - Private   

                                Industry             Sector  \
0                     Insurance Carriers          Insurance   
1                 Research & Development  Business Services   
2                             Consulting  Business Services   
3  Electrical & Electronic Manufacturing      Manufacturing   
4                Advertising & Marketing  Business Services   

                      Revenue  \
0    Unknown / Non-Applicable   
1      $1 to $2 billion (USD)   
2  $100 to $500 million (USD)   
3  $100 to $500 million (USD)   
4    Unknown / Non-Applicable   

                                         Competitors  
0            EmblemHealth, UnitedHealth Group, Aetna  
1                                                 -1  
2                                                 -1  
3  MKS Instruments, Pfeiffer Vacuum, Agilent Tech...  
4               Commerce Signals, Cardlytics, Yodlee

In [280]:
df[['Minimum_Salary', 'Maximum_Salary']] = df['Salary Estimate'].str.split('-', expand=True)

In [281]:
df.head()

index          Job Title Salary Estimate  \
0      0  Sr Data Scientist  137000-171000    
1      1     Data Scientist  137000-171000    
2      2     Data Scientist  137000-171000    
3      3     Data Scientist  137000-171000    
4      4     Data Scientist  137000-171000    

                                     Job Description  Rating  \
0  Description\n\nThe Senior Data Scientist is re...     3.1   
1  Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2  Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3  JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   

              Company Name       Location            Headquarters  \
0         Healthfirst\n3.1   New York, NY            New York, NY   
1             ManTech\n4.2  Chantilly, VA             Herndon, VA   
2      Analysis Group\n3.8     Boston, MA              Boston, MA   
3             INFICON\n3.5     Newton, MA  Bad Ragaz, Switzerland   
4  Affinity Solutions\n2.9   New York, NY            New York, NY   

                      Size  Founded        Type of ownership  \
0   1001 to 5000 employees     1993   Nonprofit Organization   
1  5001 to 10000 employees     1968         Company - Public   
2   1001 to 5000 employees     1981  Private Practice / Firm   
3    501 to 1000 employees     2000         Company - Public   
4      51 to 200 employees     1998        Company - Private   

                                Industry             Sector  \
0                     Insurance Carriers          Insurance   
1                 Research & Development  Business Services   
2                             Consulting  Business Services   
3  Electrical & Electronic Manufacturing      Manufacturing   
4                Advertising & Marketing  Business Services   

                      Revenue  \
0    Unknown / Non-Applicable   
1      $1 to $2 billion (USD)   
2  $100 to $500 million (USD)   
3  $100 to $500 million (USD)   
4    Unknown / Non-Applicable   

                                         Competitors Minimum_Salary  \
0            EmblemHealth, UnitedHealth Group, Aetna         137000   
1                                                 -1         137000   
2                                                 -1         137000   
3  MKS Instruments, Pfeiffer Vacuum, Agilent Tech...         137000   
4               Commerce Signals, Cardlytics, Yodlee         137000   

  Maximum_Salary  
0        171000   
1        171000   
2        171000   
3        171000   
4        171000

In [282]:
df['Minimum_Salary'] = pd.to_numeric(df['Minimum_Salary'], errors='coerce')
df['Maximum_Salary'] = pd.to_numeric(df['Maximum_Salary'], errors='coerce')
df['Average_Salary'] = (df['Maximum_Salary'] + df['Minimum_Salary']) / 2

In [283]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 672 entries, 0 to 671
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   index              672 non-null    int64  
 1   Job Title          672 non-null    object 
 2   Salary Estimate    672 non-null    object 
 3   Job Description    672 non-null    object 
 4   Rating             672 non-null    float64
 5   Company Name       672 non-null    object 
 6   Location           672 non-null    object 
 7   Headquarters       672 non-null    object 
 8   Size               672 non-null    object 
 9   Founded            672 non-null    int64  
 10  Type of ownership  672 non-null    object 
 11  Industry           672 non-null    object 
 12  Sector             672 non-null    object 
 13  Revenue            672 non-null    object 
 14  Competitors        672 non-null    object 
 15  Minimum_Salary     672 non-null    int64  
 16  Maximum_Salary     672 non

The new variables 'Minimum_Salary' and 'Maximum_Salary' are integers, but the 'Average_Salary' is a float, since it's the result of an average.

In [284]:
df.head()

index          Job Title Salary Estimate  \
0      0  Sr Data Scientist  137000-171000    
1      1     Data Scientist  137000-171000    
2      2     Data Scientist  137000-171000    
3      3     Data Scientist  137000-171000    
4      4     Data Scientist  137000-171000    

                                     Job Description  Rating  \
0  Description\n\nThe Senior Data Scientist is re...     3.1   
1  Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2  Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3  JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   

              Company Name       Location            Headquarters  \
0         Healthfirst\n3.1   New York, NY            New York, NY   
1             ManTech\n4.2  Chantilly, VA             Herndon, VA   
2      Analysis Group\n3.8     Boston, MA              Boston, MA   
3             INFICON\n3.5     Newton, MA  Bad Ragaz, Switzerland   
4  Affinity Solutions\n2.9   New York, NY            New York, NY   

                      Size  Founded        Type of ownership  \
0   1001 to 5000 employees     1993   Nonprofit Organization   
1  5001 to 10000 employees     1968         Company - Public   
2   1001 to 5000 employees     1981  Private Practice / Firm   
3    501 to 1000 employees     2000         Company - Public   
4      51 to 200 employees     1998        Company - Private   

                                Industry             Sector  \
0                     Insurance Carriers          Insurance   
1                 Research & Development  Business Services   
2                             Consulting  Business Services   
3  Electrical & Electronic Manufacturing      Manufacturing   
4                Advertising & Marketing  Business Services   

                      Revenue  \
0    Unknown / Non-Applicable   
1      $1 to $2 billion (USD)   
2  $100 to $500 million (USD)   
3  $100 to $500 million (USD)   
4    Unknown / Non-Applicable   

                                         Competitors  Minimum_Salary  \
0            EmblemHealth, UnitedHealth Group, Aetna          137000   
1                                                 -1          137000   
2                                                 -1          137000   
3  MKS Instruments, Pfeiffer Vacuum, Agilent Tech...          137000   
4               Commerce Signals, Cardlytics, Yodlee          137000   

   Maximum_Salary  Average_Salary  
0          171000        154000.0  
1          171000        154000.0  
2          171000        154000.0  
3          171000        154000.0  
4          171000        154000.0

In [285]:
df.describe(include='all')

index       Job Title Salary Estimate  \
count   672.000000             672             672   
unique         NaN             172              30   
top            NaN  Data Scientist   79000-131000    
freq           NaN             337              32   
mean    335.500000             NaN             NaN   
std     194.133974             NaN             NaN   
min       0.000000             NaN             NaN   
25%     167.750000             NaN             NaN   
50%     335.500000             NaN             NaN   
75%     503.250000             NaN             NaN   
max     671.000000             NaN             NaN   

                                          Job Description      Rating  \
count                                                 672  672.000000   
unique                                                489         NaN   
top     Job Overview: The Data Scientist is a key memb...         NaN   
freq                                                   12         NaN   
mean                                                  NaN    3.518601   
std                                                   NaN    1.410329   
min                                                   NaN   -1.000000   
25%                                                   NaN    3.300000   
50%                                                   NaN    3.800000   
75%                                                   NaN    4.300000   
max                                                   NaN    5.000000   

          Company Name           Location  Headquarters                 Size  \
count              672                672           672                  672   
unique             432                207           229                    9   
top     Hatch Data Inc  San Francisco, CA  New York, NY  51 to 200 employees   
freq                12                 69            33                  135   
mean               NaN                NaN           NaN                  NaN   
std                NaN                NaN           NaN                  NaN   
min                NaN                NaN           NaN                  NaN   
25%                NaN                NaN           NaN                  NaN   
50%                NaN                NaN           NaN                  NaN   
75%                NaN                NaN           NaN                  NaN   
max                NaN                NaN           NaN                  NaN   

            Founded  Type of ownership Industry                  Sector  \
count    672.000000                672      672                     672   
unique          NaN                 13       58                      23   
top             NaN  Company - Private       -1  Information Technology   
freq            NaN                397       71                     188   
mean    1635.529762                NaN      NaN                     NaN   
std      756.746640                NaN      NaN                     NaN   
min       -1.000000                NaN      NaN                     NaN   
25%     1917.750000                NaN      NaN                     NaN   
50%     1995.000000                NaN      NaN                     NaN   
75%     2009.000000                NaN      NaN                     NaN   
max     2019.000000                NaN      NaN                     NaN   

                         Revenue Competitors  Minimum_Salary  Maximum_Salary  \
count                        672         672      672.000000      672.000000   
unique                        14         108             NaN             NaN   
top     Unknown / Non-Applicable          -1             NaN             NaN   
freq                         213         501             NaN             NaN   
mean                         NaN         NaN    99196.428571   148130.952381   
std                          NaN         NaN    33009.958111    48035.110051   
min                          NaN         NaN    31000.000000    5

Now we can see that the minimum salary ranges between 31k$ to 212k$ being the average 99k$ while the maximum ranges between 56k$ and 331k$, being the average 148k$. The average salary is 123k$.

### 2) Average Salary

#### By seniority

In [286]:
df['Seniority'] = np.where(df['Job Title'].str.match('.*Senior.*|.*Sr.*|.*Expert.*|.*Experienced.*|.*Mid.*'),"Senior",
                  np.where(df['Job Title'].str.match('.*Staff.*'),"Staff",
                  np.where(df['Job Title'].str.match('.*Principal.*'),"Principal",
                  np.where(df['Job Title'].str.match('.*Manager.*|.*Director.*|.*Chief.*|.*Lead.*'),"Manager",
                  np.where(df['Job Title'].str.match('.*Early.*|.*Associate.*'),"Entry", "Entry")))))

In [287]:
df.head()

index          Job Title Salary Estimate  \
0      0  Sr Data Scientist  137000-171000    
1      1     Data Scientist  137000-171000    
2      2     Data Scientist  137000-171000    
3      3     Data Scientist  137000-171000    
4      4     Data Scientist  137000-171000    

                                     Job Description  Rating  \
0  Description\n\nThe Senior Data Scientist is re...     3.1   
1  Secure our Nation, Ignite your Future\n\nJoin ...     4.2   
2  Overview\n\n\nAnalysis Group is one of the lar...     3.8   
3  JOB DESCRIPTION:\n\nDo you have a passion for ...     3.5   
4  Data Scientist\nAffinity Solutions / Marketing...     2.9   

              Company Name       Location            Headquarters  \
0         Healthfirst\n3.1   New York, NY            New York, NY   
1             ManTech\n4.2  Chantilly, VA             Herndon, VA   
2      Analysis Group\n3.8     Boston, MA              Boston, MA   
3             INFICON\n3.5     Newton, MA  Bad Ragaz, Switzerland   
4  Affinity Solutions\n2.9   New York, NY            New York, NY   

                      Size  Founded        Type of ownership  \
0   1001 to 5000 employees     1993   Nonprofit Organization   
1  5001 to 10000 employees     1968         Company - Public   
2   1001 to 5000 employees     1981  Private Practice / Firm   
3    501 to 1000 employees     2000         Company - Public   
4      51 to 200 employees     1998        Company - Private   

                                Industry             Sector  \
0                     Insurance Carriers          Insurance   
1                 Research & Development  Business Services   
2                             Consulting  Business Services   
3  Electrical & Electronic Manufacturing      Manufacturing   
4                Advertising & Marketing  Business Services   

                      Revenue  \
0    Unknown / Non-Applicable   
1      $1 to $2 billion (USD)   
2  $100 to $500 million (USD)   
3  $100 to $500 million (USD)   
4    Unknown / Non-Applicable   

                                         Competitors  Minimum_Salary  \
0            EmblemHealth, UnitedHealth Group, Aetna          137000   
1                                                 -1          137000   
2                                                 -1          137000   
3  MKS Instruments, Pfeiffer Vacuum, Agilent Tech...          137000   
4               Commerce Signals, Cardlytics, Yodlee          137000   

   Maximum_Salary  Average_Salary Seniority  
0          171000        154000.0    Senior  
1          171000        154000.0     Entry  
2          171000        154000.0     Entry  
3          171000        154000.0     Entry  
4          171000        154000.0     Entry

In [288]:
df['Seniority'].value_counts()

Entry        558
Senior        82
Manager       16
Principal     10
Staff          6
Name: Seniority, dtype: int64

In [289]:
df_pivot = pd.pivot_table(df, values=['Average_Salary'], index = ['Seniority'], aggfunc=np.mean, sort=True)
round(df_pivot.sort_values(by=['Average_Salary'], ascending=False),2)

Average_Salary
Seniority                
Manager         140718.75
Senior          124890.24
Entry           123486.56
Staff           114750.00
Principal       101550.00

There are some  seniorities that earn more as expected, but the typical hierarchy would be Manager > Principal > Staff > Senior > Entry, but in this case we that 'Entry' level positions earn more than 'Staff' and 'Principal' due to outliers.

#### By Job Title

In [290]:
df['Job Title'].value_counts()

Data Scientist                                            337
Data Engineer                                              26
Senior Data Scientist                                      19
Machine Learning Engineer                                  16
Data Analyst                                               12
                                                         ... 
Data Science Instructor                                     1
Business Data Analyst                                       1
Purification Scientist                                      1
Data Engineer, Enterprise Analytics                         1
AI/ML - Machine Learning Scientist, Siri Understanding      1
Name: Job Title, Length: 172, dtype: int64

We can see that we have very similar positions, but written in a different way so we need to process this data:

In [291]:
df['Job Title'] = df['Job Title'].replace('Senior Data Scientist', 'Data Scientist')
df['Job Title'] = df['Job Title'].replace('Data Scientist - TS/SCI FSP or CI Required', 'Data Scientist')
df['Job Title'] = df['Job Title'].replace('Data Science Software Engineer', 'Data Scientist')
df['Job Title'] = df['Job Title'].replace('Data Science Manager, Payment Acceptance - USA', 'Data Scientist')
df['Job Title'] = df['Job Title'].replace('Geospatial Data Scientist', 'Data Scientist')
df['Job Title'] = df['Job Title'].replace('Senior Data Scientist - Algorithms', 'Data Scientist')
df['Job Title'] = df['Job Title'].replace('Developer III - Data Science', 'Data Scientist')
df['Job Title'] = df['Job Title'].replace('Staff Data Scientist - Pricing', 'Data Scientist')
df['Job Title'] = df['Job Title'].replace('Staff Data Scientist - Pricing', 'Data Scientist')
df['Job Title'] = df['Job Title'].replace(['Environmental Data Science', 'Director of Data Science', 'Data Scientist - Statistics, Mid-Career', 'Data Science Instructor', 'Applied Technology Researcher / Data Scientist', 'Lead Data Scientist – Network Analysis and Control'
                                           , 'Applied AI Scientist / Engineer','Data Science All Star Program - Data Engineer Track','Software Engineer (Data Scientist, C,C++,Linux,Unix) - SISW - MG', 'Chief Scientist','Data Scientist Machine Learning',
                                           'Senior Clinical Data Scientist Programmer','Senior Data & Machine Learning Scientist','Aviation AI/ML Data Scientist','Manager / Lead, Data Science & Analytics',
                                           'AI Data Scientist','Staff Data Scientist','Health Data Scientist - Biomedical/Biostats','Statistical Scientist','Big Data Engineer','Software Engineer - Data Science',
                                           'Data Scientist (TS/SCI)','Data Science Manager','Data Scientist Technical Specialist','Data Science Analyst','Software Engineer - Machine Learning & Data Science (Applied Intelligence Services Team)',
                                           'Senior Principal Data Scientist (Python/R)','Data Scientist- Industrial Discrete Sector Industry','Data Scientist, Kinship - NYC/Portland','Data Scientist(s)/Machine Learning Engineer',
                                           'Principal Data Scientist - Machine Learning','(Sr.) Data Scientist -','Patient Safety- Associate Data Scientist','Senior Data Scientist - R&D Oncology','Data Scientist / Applied Mathematician',
                                           'Data Scientist - Image and Video Analytics','Real World Science, Data Scientist','Data Scientist 3 (718)','Data Scientist, Applied Machine Learning - Bay Area','Data Scientist - Machine Learning',
                                           'IT Partner Digital Health Technology and Data Science','Computer Vision / Deep Learning Scientist','Data Scientist (TS/SCI w/ Poly)','Applied Computer Scientist',
                                           'Principal Data Scientist','Lead Data Scientist','Sr. ML/Data Scientist - AI/NLP/Chatbot','AI Ops Data Scientist','Data Scientist - TS/SCI Required','Senior Data Scientist – Image Analytics, Novartis AI Innovation Lab',
                                           'Decision Scientist','Data Scientist / Machine Learning Expert','Staff Data Scientist - Analytics','Data Scientist - Statistics, Early Career','Data Scientist-Human Resources','Senior Research Statistician- Data Scientist',
                                           'Associate Data Scientist','Data Scientist - Intermediate','Sr. Data Scientist','VP, Data Science','Product Data Scientist - Ads Data Science','Sr. Data Scientist II',
                                           'Data & Machine Learning Scientist','Sr Data Scientist','Data Scientist - Risk','Data Scientist/Machine Learning','Data Scientist - Contract','Experienced Data Scientist','Computer Scientist 1','Computational Scientist'],'Data Scientist')



df['Job Title'] = df['Job Title'].replace('Operations Data Analyst', 'Data Analyst')
df['Job Title'] = df['Job Title'].replace('Sr Data Analyst', 'Data Analyst')
df['Job Title'] = df['Job Title'].replace('Report Writer-Data Analyst', 'Data Analyst')
df['Job Title'] = df['Job Title'].replace('Say Business Data Analyst', 'Data Analyst')
df['Job Title'] = df['Job Title'].replace('Senior Data Analyst', 'Data Analyst')
df['Job Title'] = df['Job Title'].replace('Business Data Analyst', 'Data Analyst')
df['Job Title'] = df['Job Title'].replace('Global Data Analyst', 'Data Analyst')
df['Job Title'] = df['Job Title'].replace('E-Commerce Data Analyst', 'Data Analyst')
df['Job Title'] = df['Job Title'].replace(['Data Scientist/Data Analytics Practitioner','Data Analyst I','Enterprise Data Analyst (Enterprise Portfolio Management Office)','Market Research Data Scientist','Business Intelligence Analyst I- Data Insights',
                                           'Analytics Manager','Diversity and Inclusion Data Analyst','Jr. Business Data Analyst (position added 6/12/2020)','Production Engineer - Statistics/Data Analysis',
                                           'Data Analyst II','RFP Data Analyst','Sr. Data Analyst','Data Analyst/Engineer','Clinical Data Analyst','Senior Data Analyst - Finance & Platform Analytics',
                                           'In-Line Inspection Data Analyst','Senior Business Intelligence Analyst','Analytics - Business Assurance Data Analyst','Business Intelligence Analyst','Senior Analyst/Data Scientist',
                                           'Data Analyst - Unilever Prestige','Intelligence Data Analyst, Senior','Analytics Manager - Data Mart','Health Plan Data Analyst, Sr','Say Business Data Analyst',
                                           'Sr Data Analyst'], 'Data Analyst')


df['Job Title'] = df['Job Title'].replace('Software Data Engineer', 'Data Engineer')
df['Job Title'] = df['Job Title'].replace('Data Engineer, Enterprise Analytics', 'Data Engineer')
df['Job Title'] = df['Job Title'].replace('Jr. Data Engineer', 'Data Engineer')
df['Job Title'] = df['Job Title'].replace('Senior Data Engineer', 'Data Engineer')
df['Job Title'] = df['Job Title'].replace('Data Engineer (Remote)', 'Data Engineer')
df['Job Title'] = df['Job Title'].replace(['Staff BI and Data Engineer','Data Integration and Modeling Engineer','Principal Data & Analytics Platform Engineer','Information Systems Engineering Specialist (Engineering Scientist)','Data Solutions Engineer - Data Modeler'
                                           ,'Sr Data Engineer (Sr BI Developer)','Data Engineer, Digital & Comp Pathology','Tableau Data Engineer 20-0117','Cloud Data Engineer (Azure)','Data Engineer - Kafka','Data Modeler (Analytical Systems)',
                                           'Data Engineer (Analytics, SQL, Python, AWS)','Data Analytics Engineer','Data Architect'], 'Data Engineer')



df['Job Title'] = df['Job Title'].replace('Machine Learning Engineer ', 'Machine Learning Engineer')
df['Job Title'] = df['Job Title'].replace('AI/ML - Machine Learning Scientist, Siri Understanding', 'Machine Learning Engineer')
df['Job Title'] = df['Job Title'].replace('Machine Learning Scientist - Bay Area, CA', 'Machine Learning Engineer')
df['Job Title'] = df['Job Title'].replace(['Computational Scientist, Machine Learning','Machine Learning Engineer, Sr.','Machine Learning Scientist / Engineer','Principal Machine Learning Scientist','Senior Machine Learning Engineer',
                                           'Senior Machine Learning Scientist - Bay Area, CA','Scientist - Machine Learning','Machine Learning Engineer/Scientist','AI/ML - Machine Learning Scientist, Siri Understanding'], 'Machine Learning Engineer')


df['Job Title'] = df['Job Title'].replace('ENGINEER - COMPUTER SCIENTIST - RESEARCH COMPUTER SCIENTIST - SIGNAL PROCESSING - SAN ANTONIO OR', 'Other')
df['Job Title'] = df['Job Title'].replace('Hydrogen/Tritium Materials Scientist (Experienced)', 'Other')
df['Job Title'] = df['Job Title'].replace('ELISA RESEARCH SCIENTIST (CV-15)', 'Other')
df['Job Title'] = df['Job Title'].replace('Computational Scientist', 'Other')
df['Job Title'] = df['Job Title'].replace('Equity Data Insights Analyst - Quantitative Analyst', 'Other')
df['Job Title'] = df['Job Title'].replace(['Purification Scientist', 'Real World Evidence (RWE) Scientist', 'Sr. Research Associate/ Scientist, NGS prep & Molecular Genomics','NGS Scientist',
                                           'Vice President, Biometrics and Clinical Data Management','Scientist - Molecular Biology','Scientist/Research Associate-Metabolic Engineering','Human Factors Scientist','Research Scientist Patient Preferences (Remote)',
                                           'Development Scientist, Voltaren','Staff Scientist- Upstream PD','Sr Scientist - Extractables & Leachables','Medical Lab Scientist',
                                           'Lead Certified Clinical Laboratory Scientist - Saturday - Tuesday, 8:00pm - 6:30am shift','COMPUTER SCIENTIST - ENGINEER - RESEARCH COMPUTER SCIENTIST - TRANSPORTATION TECHNOLOGY',
                                           'COMPUTER SCIENTIST - ENGINEER - RESEARCH COMPUTER SCIENTIST - SIGNAL PROCESSING','Manager, Field Application Scientist, Southeast','Scientist - Biomarker and Flow Cytometry','Data Modeler',
                                           'Scientist / Group Lead, Cancer Biology','Computational Behavioral Scientist','Research Scientist - Patient-Centered Research (Remote)','Senior Scientist - Toxicologist - Product Integrity (Stewardship)',
                                           'Weapons and Sensors Engineer/Scientist','Principal Scientist/Associate Director, Quality Control and Analytical Technologies'], 'Other')

In [292]:
df_pivot = pd.pivot_table(df, values=['Average_Salary'], index = ['Job Title'], aggfunc=np.mean, sort=True)
round(df_pivot.sort_values(by=['Average_Salary'], ascending=False),2)

Average_Salary
Job Title                                
Other                           137276.32
Data Scientist                  125077.16
Machine Learning Engineer       117953.12
Data Analyst                    116352.46
Data Engineer                   113200.00

From this table we can see that the 'Other' positions are actually the ones that earn the most, because of roles such as 'Manager, Field Application Scientist' or 'Computer Scientist - Engineer - Research Computer Scientist - Signal Processing' which have a very high Average Salary which are definitely outliers.

From the other jobs, Data Scientists earn more than ML Engineers which in turn earns more than a Data Analyst and the ones with the lowest average salary are Data Engineers.

#### By Sector

In [293]:
df_pivot_industry = pd.pivot_table(df, values=['Average_Salary'], index = ['Sector'], aggfunc=np.mean, sort=True)
round(df_pivot_industry.sort_values(by=['Average_Salary'], ascending=False),2)

Average_Salary
Sector                                            
Consumer Services                        203750.00
Media                                    155300.00
Retail                                   150142.86
Government                               134470.59
Aerospace & Defense                      132695.65
Business Services                        130079.17
-1                                       129669.01
Travel & Tourism                         128666.67
Accounting & Legal                       124666.67
Manufacturing                            123239.13
Biotech & Pharmaceuticals                122871.21
Health Care                              119761.90
Information Technology                   118909.57
Education                                117833.33
Transportation & Logistics               117750.00
Real Estate                              116833.33
Finance                                  115500.00
Telecommunications                       112357.14
Insurance                                111281.25
Agriculture & Forestry                   103833.33
Oil, Gas, Energy & Utilities             101300.00
Construction, Repair & Maintenance        74750.00
Non-Profit                                43500.00

By Sector, it looks like 'Consumer Services' workers get the most money, followed by Media and Retail. On the other extreme, Non Profit, Construction, Repair & Maintenance and Oil, Gas, Energy & Utilities earn the least.

### 3) Creating State

In [294]:
df['State'] = np.where(df['Location'].str.match('.*,.*'),df['Location'].str[-2:],"Other")

In [295]:
df['State'].value_counts()

CA       165
VA        89
MA        62
NY        52
MD        40
IL        30
DC        26
Other     23
TX        17
WA        16
OH        14
MO        12
PA        12
CO        10
NJ        10
NC         9
GA         9
FL         8
TN         8
OK         6
WI         6
IN         5
MI         5
AZ         4
AL         4
MN         4
CT         4
UT         3
NE         3
IA         3
OR         2
SC         2
RI         2
NH         2
LA         1
MS         1
KS         1
DE         1
WV         1
Name: State, dtype: int64

In [296]:
df[df['State'] == "Other"]

index                  Job Title Salary Estimate  \
16      16             Data Scientist  137000-171000    
34      34              Data Engineer   75000-131000    
53      53              Data Engineer   75000-131000    
89      89             Data Scientist   79000-131000    
93      93             Data Scientist   79000-131000    
100    100             Data Scientist   99000-132000    
134    134  Machine Learning Engineer   90000-109000    
135    135  Machine Learning Engineer   90000-109000    
199    199             Data Scientist   79000-106000    
227    227             Data Scientist   71000-123000    
237    237             Data Scientist   71000-123000    
305    305             Data Scientist   145000-225000   
307    307             Data Scientist   145000-225000   
438    438             Data Scientist   79000-133000    
473    473             Data Scientist    31000-56000    
483    483             Data Scientist    31000-56000    
510    510             Data Scientist  212000-331000    
570    570             Data Scientist  138000-158000    
600    600             Data Scientist   80000-132000    
601    601             Data Scientist   80000-132000    
611    611             Data Scientist   87000-141000    
620    620             Data Scientist   87000-141000    
632    632             Data Scientist   92000-155000    

                                       Job Description  Rating  \
16   We are an ambitious, well-funded startup with ...     4.1   
34   Chef Software is the industry leader in IT aut...     3.6   
53   Jr. Data Engineer\n\nJob Details\nLevel\nEntry...     4.8   
89   Back to search results\nPrevious job\n\nNext j...     3.9   
93   Thursday, June 11, 2020\n\nMerrick Bank is a t...     3.6   
100  Job Description\n\n\nAvlino Inc. is seeking ex...     4.9   
134  Role Description\nTriplebyte screens and evalu...     3.2   
135  Role Description\nTriplebyte screens and evalu...     3.2   
199  We are an ambitious, well-funded startup with ...     4.1   
227  About The Position\n\nREE is redefining how mo...     5.0   
237  Who we are\n\n\nFueled by a fundamental belief...     3.8   
305  Why TrueAccord?\n\nDebt collection is failing ...     3.4   
307  DATA SCIENTIST REMOTE working with occasional ...     3.2   
438  Title Data Scientist Location Remote Duration ...    -1.0   
473  Data Scientist\n\nTrexquant is a systematic he...     4.0   
483  Company overview:\nFor more than three decades...     2.9   
510  Position: Data Scientist\nLocation: Denver\nSt...     3.6   
570  Back to search results\nPrevious job\n\nNext j...     3.9   
600  Company overview:\nFor more than three decades...     2.9   
601  ConsumerTrack™ is unique in the digital market...     3.2   
611  Thursday, June 11, 2020\n\nMerrick Bank is a t...     3.6   
620  Business Unit:\nCubic Transportation Systems\n...     3.3   
632  Job Description\n\n\nAvlino Inc. is seeking ex...     4.9   

                    Company Name       Location           Headquarters  \
16   Upside Business Travel\n4.1         Remote         Washington, DC   
34                     Chef\n3.6  United States            Seattle, WA   
53              IT Concepts\n4.8         Remote             Vienna, VA   
89                     Mars\n3.9  United States            Mc Lean, VA   
93             Merrick Bank\n3.6           Utah       South Jordan, UT   
100                  Avlino\n4.9     New Jersey            Holmdel, NJ   
134              Triplebyte\n3.2         Remote      San Francisco, CA   
135              Triplebyte\n3.2         Remote      San Francisco, CA   
199  Upside Business Travel\n4.1         Remote         Washington, DC   
227                     REE\n5.0  United States  Tel Aviv-Yafo, Israel   
237                  PayPal\n3.8          Texas           San Jose, CA   
305              TrueAccord\n3.4         Remote      San Francisco, CA   
307          Kollasoft Inc.\n3.2  United States         Scottsdale, AZ   
438          Bl

We can still transform into state 'New Jersey', 'Utah', 'Texas' and 'California'

In [297]:
df['State'] = np.where(df['Location'].str.match('.*New Jersey.*'),"NJ",
                       np.where(df['Location'].str.match('.*Utah.*'),"UT",
                                np.where(df['Location'].str.match('.*Texas.*'),"TX",
                                         np.where(df['Location'].str.match('.*California.*'),"CA",df['State']))))

In [298]:
df[df['State'] == "Other"]

index                  Job Title Salary Estimate  \
16      16             Data Scientist  137000-171000    
34      34              Data Engineer   75000-131000    
53      53              Data Engineer   75000-131000    
89      89             Data Scientist   79000-131000    
134    134  Machine Learning Engineer   90000-109000    
135    135  Machine Learning Engineer   90000-109000    
199    199             Data Scientist   79000-106000    
227    227             Data Scientist   71000-123000    
305    305             Data Scientist   145000-225000   
307    307             Data Scientist   145000-225000   
438    438             Data Scientist   79000-133000    
473    473             Data Scientist    31000-56000    
483    483             Data Scientist    31000-56000    
510    510             Data Scientist  212000-331000    
570    570             Data Scientist  138000-158000    
600    600             Data Scientist   80000-132000    
620    620             Data Scientist   87000-141000    

                                       Job Description  Rating  \
16   We are an ambitious, well-funded startup with ...     4.1   
34   Chef Software is the industry leader in IT aut...     3.6   
53   Jr. Data Engineer\n\nJob Details\nLevel\nEntry...     4.8   
89   Back to search results\nPrevious job\n\nNext j...     3.9   
134  Role Description\nTriplebyte screens and evalu...     3.2   
135  Role Description\nTriplebyte screens and evalu...     3.2   
199  We are an ambitious, well-funded startup with ...     4.1   
227  About The Position\n\nREE is redefining how mo...     5.0   
305  Why TrueAccord?\n\nDebt collection is failing ...     3.4   
307  DATA SCIENTIST REMOTE working with occasional ...     3.2   
438  Title Data Scientist Location Remote Duration ...    -1.0   
473  Data Scientist\n\nTrexquant is a systematic he...     4.0   
483  Company overview:\nFor more than three decades...     2.9   
510  Position: Data Scientist\nLocation: Denver\nSt...     3.6   
570  Back to search results\nPrevious job\n\nNext j...     3.9   
600  Company overview:\nFor more than three decades...     2.9   
620  Business Unit:\nCubic Transportation Systems\n...     3.3   

                    Company Name       Location           Headquarters  \
16   Upside Business Travel\n4.1         Remote         Washington, DC   
34                     Chef\n3.6  United States            Seattle, WA   
53              IT Concepts\n4.8         Remote             Vienna, VA   
89                     Mars\n3.9  United States            Mc Lean, VA   
134              Triplebyte\n3.2         Remote      San Francisco, CA   
135              Triplebyte\n3.2         Remote      San Francisco, CA   
199  Upside Business Travel\n4.1         Remote         Washington, DC   
227                     REE\n5.0  United States  Tel Aviv-Yafo, Israel   
305              TrueAccord\n3.4         Remote      San Francisco, CA   
307          Kollasoft Inc.\n3.2  United States         Scottsdale, AZ   
438          Blue Icy Water, LLC  United States            Livonia, MI   
473    Trexquant Investment\n4.0  United States           Stamford, CT   
483        Pragmatics, Inc.\n2.9  United States             Reston, VA   
510         Creative Circle\n3.6  United States        Los Angeles, CA   
570                    Mars\n3.9  United States            Mc Lean, VA   
600        Pragmatics, Inc.\n2.9  United States             Reston, VA   
620                   Cubic\n3.3  United States          San Diego, CA   

                        Size  Founded  Type of ownership  \
16       51 to 200 employees     2015  Company - Private   
34      201 to 500 employees     2008  Company - Private   
53       51 to 200 employees     2003  Company - Private   
89          10000+ employees     1911  Company - Private   
134      51 to 200 employees     2015  Company - Private   
135      51 to 200 employees     2015  Company - Private   
199      51 to 200 employees     201

In [299]:
df['State'].value_counts()

CA       166
VA        89
MA        62
NY        52
MD        40
IL        30
DC        26
TX        18
Other     17
WA        16
OH        14
NJ        12
PA        12
MO        12
CO        10
GA         9
NC         9
FL         8
TN         8
OK         6
WI         6
UT         5
MI         5
IN         5
CT         4
MN         4
AL         4
AZ         4
IA         3
NE         3
SC         2
OR         2
RI         2
NH         2
MS         1
LA         1
KS         1
DE         1
WV         1
Name: State, dtype: int64

In [300]:
df_pivot = pd.pivot_table(df, values=['Average_Salary'], index = ['State'], aggfunc=np.mean, sort=True)
round(df_pivot.sort_values(by=['Average_Salary'], ascending=False),2)

Average_Salary
State                
DE          271500.00
NC          150111.11
WI          144416.67
AZ          140875.00
IA          140833.33
DC          139500.00
NY          136432.69
WA          134781.25
TX          133888.89
MS          133000.00
OR          132000.00
NH          132000.00
AL          131125.00
RI          130750.00
VA          126769.66
PA          125833.33
OK          122916.67
MA          121967.74
OH          121714.29
Other       120941.18
IL          120883.33
CA          120521.08
NJ          120416.67
MI          119000.00
FL          118437.50
MO          114958.33
WV          114000.00
IN          113400.00
MD          112375.00
GA          109666.67
NE          107666.67
UT          106700.00
CO          103650.00
KS          103000.00
LA          103000.00
TN          102187.50
CT           98125.00
SC           96000.00
MN           94000.00

From what we can see here, Delaware (DE) has a vastly higher Average Salary in comparison to all the other states. On the other side, Minnesota (MN) has the lowest salary.

In [301]:
df[df['State'] == "DE"]

index       Job Title Salary Estimate  \
509    509  Data Scientist  212000-331000    

                                       Job Description  Rating  \
509  Title: Real World Science, Data Scientist\nLoc...     4.0   

         Company Name        Location               Headquarters  \
509  AstraZeneca\n4.0  Wilmington, DE  Cambridge, United Kingdom   

                 Size  Founded Type of ownership                   Industry  \
509  10000+ employees     1913  Company - Public  Biotech & Pharmaceuticals   

                        Sector             Revenue  \
509  Biotech & Pharmaceuticals  $10+ billion (USD)   

                          Competitors  Minimum_Salary  Maximum_Salary  \
509  Roche, GlaxoSmithKline, Novartis          212000          331000   

     Average_Salary Seniority State  
509        271500.0     Entry    DE

As we can see, Delaware only has one job posting with extremely high salary.

### 4) Skills

AWS
Azure
Hadoop
Python
Tensorflow
Pytorch
NoSQL
PostgreSQL
SAS
Java
Scala
Spark
Excel
Tableau
Big Data

In [302]:
df['AWS'] = df['Job Description'].str.contains('AWS', case=False).astype(int)
df['Azure'] = df['Job Description'].str.contains('Azure', case=False).astype(int)
df['Python'] = df['Job Description'].str.contains('Python', case=False).astype(int)
df['Tensorflow'] = df['Job Description'].str.contains('Tensorflow', case=False).astype(int)
df['Pytorch'] = df['Job Description'].str.contains('Pytorch', case=False).astype(int)
df['NoSQL'] = df['Job Description'].str.contains('NoSQL', case=False).astype(int)
df['PostgreSQL'] = df['Job Description'].str.contains('PostgreSQL', case=False).astype(int)
df['SAS'] = df['Job Description'].str.contains('SAS', case=False).astype(int)
df['Java'] = df['Job Description'].str.contains('Java', case=False).astype(int)
df['Scala'] = df['Job Description'].str.contains('Scala', case=False).astype(int)
df['Spark'] = df['Job Description'].str.contains('Spark', case=False).astype(int)
df['Excel'] = df['Job Description'].str.contains('Excel', case=False).astype(int)
df['Tableau'] = df['Job Description'].str.contains('Tableau', case=False).astype(int)
df['Big_Data'] = df['Job Description'].str.contains('Big Data', case=False).astype(int)

In [303]:
df.describe(include=None)

index      Rating      Founded  Minimum_Salary  Maximum_Salary  \
count  672.000000  672.000000   672.000000      672.000000      672.000000   
mean   335.500000    3.518601  1635.529762    99196.428571   148130.952381   
std    194.133974    1.410329   756.746640    33009.958111    48035.110051   
min      0.000000   -1.000000    -1.000000    31000.000000    56000.000000   
25%    167.750000    3.300000  1917.750000    79000.000000   119000.000000   
50%    335.500000    3.800000  1995.000000    91000.000000   133000.000000   
75%    503.250000    4.300000  2009.000000   122000.000000   165000.000000   
max    671.000000    5.000000  2019.000000   212000.000000   331000.000000   

       Average_Salary         AWS       Azure      Python  Tensorflow  \
count      672.000000  672.000000  672.000000  672.000000  672.000000   
mean    123663.690476    0.258929    0.074405    0.730655    0.153274   
std      39580.267895    0.438373    0.262624    0.443950    0.360520   
min      43500.000000    0.000000    0.000000    0.000000    0.000000   
25%     103000.000000    0.000000    0.000000    0.000000    0.000000   
50%     114000.000000    0.000000    0.000000    1.000000    0.000000   
75%     136500.000000    1.000000    0.000000    1.000000    0.000000   
max     271500.000000    1.000000    1.000000    1.000000    1.000000   

          Pytorch       NoSQL  PostgreSQL         SAS        Java       Scala  \
count  672.000000  672.000000  672.000000  672.000000  672.000000  672.000000   
mean     0.081845    0.114583    0.028274    0.144345    0.247024    0.248512   
std      0.274333    0.318756    0.165878    0.351701    0.431602    0.432472   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

            Spark       Excel     Tableau    Big_Data  
count  672.000000  672.000000  672.000000  672.000000  
mean     0.281250    0.440476    0.183036    0.206845  
std      0.449944    0.496814    0.386984    0.405345  
min      0.000000    0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000    0.000000  
50%      0.000000    0.000000    0.000000    0.000000  
75%      1.000000    1.000000    0.000000    0.000000  
max      1.000000    1.000000    1.000000    1.000000

Now we know that technologies/programming languages that are referred to the most are: Python, Excel and Spark and the least are Azure, Pytorch and PostgreSQL